app/server/lib/OIDCConfig.ts

```diff
@@ -33,42 +36,45 @@
  */
 
 import * as express from 'express';
-import {GristLoginSystem, GristServer} from './GristServer';
+import { GristLoginSystem, GristServer } from './GristServer';
 import { Client, generators, Issuer, UserinfoResponse } from 'openid-client';
-import {Sessions} from './Sessions';
+import { Sessions } from './Sessions';
 import log from 'app/server/lib/log';
-import {Permit} from './Permit';
+import { appSettings } from './AppSettings';
+import { RequestWithLogin } from './Authorizer';
 
 const CALLBACK_URL = '/oauth2/callback';
 
-// Variables used for temporary permits.
-const LOGIN_ACTION = 'oidc-login';
-const WAIT_IN_MINUTES = 20;
-
 export class OIDCConfig {
   private _client: Client;
   private _redirectUrl: string;
 
-  public constructor(private _gristServer: GristServer) {
+  public constructor() {
   }
 
   public async initOIDC(): Promise<void> {
-    if (!process.env.GRIST_OIDC_SP_HOST) { throw new Error('initOIDC requires GRIST_OIDC_SP_HOST to be set'); }
-    if (!process.env.GRIST_OIDC_IDP_ISSUER) { throw new Error('initOIDC requires GRIST_OIDC_IDP_ISSUER to be set'); }
-    if (!process.env.GRIST_OIDC_IDP_CLIENT_ID) {
-      throw new Error('initOIDC requires GRIST_OIDC_IDP_CLIENT_ID to be set');
-    }
-    if (!process.env.GRIST_OIDC_IDP_CLIENT_SECRET) {
-      throw new Error('initOIDC requires GRIST_OIDC_IDP_CLIENT_SECRET to be set');
-    }
-    const spHost: string = process.env.GRIST_OIDC_SP_HOST;
-    const issuer = await Issuer.discover(process.env.GRIST_OIDC_IDP_ISSUER);
+    const section = appSettings.section('login').section('system').section('oidc');
+    const spHost = section.flag('spHost').requireString({
+      envVar: 'GRIST_OIDC_SP_HOST',
+      defaultValue: process.env.APP_HOME_URL,
+    });
+    const issuerUrl = section.flag('issuer').requireString({
+      envVar: 'GRIST_OIDC_IDP_ISSUER',
+    });
+    const clientId = section.flag('clientId').requireString({
+      envVar: 'GRIST_OIDC_IDP_CLIENT_ID',
+    });
+    const clientSecret = section.flag('clientSecret').requireString({
+      envVar: 'GRIST_OIDC_IDP_CLIENT_SECRET',
```

Comment:
There is a handy `censor: true` setting you can pass, to ensure the value doesn't get logged as plaintext during server startup.